tt_metal/detail/tt_metal.hpp

```diff
@@ -267,18 +267,35 @@ namespace tt::tt_metal{
             return true;
         }
 
-        inline CommandQueue &GetCommandQueue(Device *device)
+
+
+        inline CommandQueue& GetCommandQueue ( Device * device)
         {
             detail::DispatchStateCheck(true);
-            // For now there is only one SW CommandQueue per device
             static std::vector<std::unique_ptr<CommandQueue>> command_queues( GetNumAvailableDevices() );
+            static std::mutex cq_creation_mutex;
+            chip_id_t id = device->id();
+            {
+                std::lock_guard<std::mutex> lock(cq_creation_mutex);
+                if ( command_queues[device->id()] == nullptr || command_queues[device->id()]->device() != device)
+                    command_queues[device->id()] = std::make_unique<CommandQueue>( device, 0);
+            }
+            return *(command_queues[id]);
+        }
+
+        inline HWCommandQueue &GetHWCommandQueue(Device *device, uint32_t cmd_queue_channel)
+        {
+            detail::DispatchStateCheck(true);
+            // For now there is only one SW HWCommandQueue per device
+            static std::vector<std::unique_ptr<HWCommandQueue>> command_queues( GetNumAvailableDevices() );
```

Comment:
_CommandQueue_ == SW command queue
_HWCommandQueue_ == Object for mananing hardware command queue

Currently,
- there is only one CommandQueue for each device
- there is only one HWCommandQueue for each device, even though there can be multiple command queue channels

Am I correct?

Reply:
If the above are correct, what I remember is that Metal team will implement SW command queues in the ttDNN layer. However, it seems like SW command queues are implemented in tt_metal layer.

Actually, I do not matter whether the SW command queues are implemented in the ttDNN layer or tt_metal layer.
I am just asking to clearly understand what is Metal team's plan.

Reply:
Yes I plan on having the `Queue` in ttDNN layer be simply an alias for `CommandQueue` in metal layer

Reply:
There can be up to 2 `HWCommandQueues` per device. Right now, the plan is to support a 1:1 mapping between `CommandQueue` and `HWCommandQueue`, but I personally don't like these limitations, and will hopefully be able to support k:1 mappings between `CommandQueue` and `HWCommandQueue`